src/actor/apply.rs

```diff
@@ -0,0 +1,155 @@
+use super::calc;
+use super::status_effect::status::{SetCombo, Status, StatusFlag, StatusFlags};
+use super::status_effect::StatusEffect;
+use super::{ActiveCombos, QueryActor};
+use crate::sim::{SimState, SimTime};
+use bevy_ecs::prelude::Entity;
+use std::sync::Arc;
+
+pub trait Apply {
+    fn apply(&self, sim: &SimState, query: &mut QueryActor, source: Entity, target: Entity);
+}
+
+#[derive(Default)]
+pub struct DoDirectDamage {
+    pub potency: i64,
+    pub combo_action_id: Option<u32>,
+    pub combo_potency: Option<i64>,
+    pub attack_type: calc::AttackType,
+}
+
+impl DoDirectDamage {
+    fn consume_combo(&self, active_combos: &mut ActiveCombos) -> bool {
+        if let Some(action_id) = self.combo_action_id {
+            if active_combos.has_action(&action_id) {
+                active_combos.remove_action(&action_id);
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl Apply for DoDirectDamage {
+    fn apply(&self, sim: &SimState, query: &mut QueryActor, source: Entity, target: Entity) {
+        let calculated_damage;
+        if let Ok((_, job, _, _, _, _, mut status_effects, stats, mut active_combos)) =
+            query.get_mut(source)
+        {
+            let mut potency = self.potency;
+            if self.consume_combo(&mut active_combos) {
+                potency = match self.combo_potency {
+                    Some(combo_potency) => combo_potency,
+                    None => panic!(
+                        "Consumed a combo, but no combo_potency is set. This should not happen."
+                    ),
+                }
+            }
+            calculated_damage =
+                calc::direct_damage(sim, potency, *job, &stats, self.attack_type, vec![]);
+            status_effects.expire_with_flag(StatusFlag::ExpireOnDirectDamage);
+        } else {
+            panic!("Tried to get stats of a source with no stats.")
+        }
+
+        if let Ok((_, _, _, _, _, mut damage, _, _, _)) = query.get_mut(target) {
+            damage.add(calculated_damage);
+        } else {
+            panic!("Tried to do damage to a target that has no Damage component.")
+        }
+    }
+}
+
+pub struct StartRecast {
+    pub action_id: u32,
+    pub duration: SimTime,
+}
+
+impl Apply for StartRecast {
+    fn apply(&self, sim: &SimState, query: &mut QueryActor, source: Entity, _target: Entity) {
+        if let Ok((_, _, _, _, mut recast_expirations, _, _, _, _)) = query.get_mut(source) {
+            recast_expirations.set(self.action_id, sim.now() + self.duration);
+        }
+    }
+}
+
+pub struct GiveStatusEffect {
+    pub status: Status,
+    pub target_source: bool,

```

Comment:
This is a bit surprising that it's a bool?

Reply:
It's supposed to say "target the source instead of the target". The receiver of the status effect is set on 83 depending on this bool. There's probably a better name for it, as I can see what you mean that it's unclear in isolation.